Mathlib/Analysis/SpecialFunctions/Log/ENNReal.lean

```diff
@@ -0,0 +1,231 @@
+Conventions about multiplication in `ℝ≥0∞` and addition in `EReal` make the identity
+`log (x * y) = log x + log y` unconditional. -/
+noncomputable def log (x : ℝ≥0∞) : EReal :=
+  if x = 0 then ⊥
+    else if x = ⊤ then ⊤
+    else Real.log (ENNReal.toReal x)
+
+@[simp]
+theorem log_zero : log 0 = ⊥ := by simp [log]
+
+@[simp]
+theorem log_one : log 1 = 0 := by simp [log]
+
+@[simp]
+theorem log_top : log ⊤ = ⊤ := by simp [log]
+
+theorem log_pos_real {x : ℝ≥0∞} (h : x ≠ 0) (h' : x ≠ ⊤) :
+    log x = Real.log (ENNReal.toReal x) := by simp [log, h, h']
+
+theorem log_pos_real' {x : ℝ≥0∞} (h : 0 < x.toReal) :
+    log x = Real.log (ENNReal.toReal x) := by
+  simp [log, Ne.symm (ne_of_lt (ENNReal.toReal_pos_iff.1 h).1),
+    ne_of_lt (ENNReal.toReal_pos_iff.1 h).2]
+
+theorem log_of_nnreal {x : ℝ≥0} (h : x ≠ 0) :
+    log (x : ℝ≥0∞) = Real.log x := by simp [log, h]
+
+end Definition
+
+/-! ### Monotonicity -/
+section Monotonicity
+
+theorem log_strictMono : StrictMono log := by
+  intro x y h
+  unfold log
+  rcases ENNReal.trichotomy x with (rfl | rfl | x_real)
+  · rcases ENNReal.trichotomy y with (rfl | rfl | y_real)
+    · exfalso; exact lt_irrefl 0 h
+    · simp
+    · simp [Ne.symm (ne_of_lt (ENNReal.toReal_pos_iff.1 y_real).1),
+      ne_of_lt (ENNReal.toReal_pos_iff.1 y_real).2, EReal.bot_lt_coe]
+  · exfalso; exact (ne_top_of_lt h) (Eq.refl ⊤)
+  · simp only [Ne.symm (ne_of_lt (ENNReal.toReal_pos_iff.1 x_real).1),
+      ne_of_lt (ENNReal.toReal_pos_iff.1 x_real).2]
+    rcases ENNReal.trichotomy y with (rfl | rfl | y_real)
+    · exfalso; rw [← ENNReal.bot_eq_zero] at h; exact not_lt_bot h
+    · simp
+    · simp only [Ne.symm (ne_of_lt (ENNReal.toReal_pos_iff.1 y_real).1), ↓reduceIte,
+        ne_of_lt (ENNReal.toReal_pos_iff.1 y_real).2, EReal.coe_lt_coe_iff]
+      apply Real.log_lt_log x_real
+      exact (ENNReal.toReal_lt_toReal (ne_of_lt (ENNReal.toReal_pos_iff.1 x_real).2)
+        (ne_of_lt (ENNReal.toReal_pos_iff.1 y_real).2)).2 h
+
+theorem log_monotone : Monotone log := StrictMono.monotone log_strictMono
+
+theorem log_injective : Function.Injective log := StrictMono.injective log_strictMono
+
+theorem log_surjective : Function.Surjective log := by
+  intro y
+  cases' eq_bot_or_bot_lt y with y_bot y_nbot
+  · exact y_bot ▸ ⟨⊥, log_zero⟩
+  cases' eq_top_or_lt_top y with y_top y_ntop
+  · exact y_top ▸ ⟨⊤, log_top⟩
+  use ENNReal.ofReal (Real.exp y.toReal)
+  have exp_y_pos := not_le_of_lt (Real.exp_pos y.toReal)
+  simp only [log, ofReal_eq_zero, exp_y_pos, ↓reduceIte, ofReal_ne_top,
+    ENNReal.toReal_ofReal (le_of_lt (Real.exp_pos y.toReal)), Real.log_exp y.toReal]
+  exact EReal.coe_toReal (ne_of_lt y_ntop) (Ne.symm (ne_of_lt y_nbot))
+
+theorem log_bijective : Function.Bijective log := ⟨log_injective, log_surjective⟩
+
+/-- `log` as an order isomorphism. -/
+noncomputable def log_orderIso : ℝ≥0∞ ≃o EReal :=
+  StrictMono.orderIsoOfSurjective log log_strictMono log_surjective
+
```

Comment:
Could you add a simp lemma saying that `log_orderIso x = log x`?

Reply:
Done